### PR TITLE
scheduler tests 1

### PR DESCRIPTION
**Pull request by cgardens**
## What
* Add `Jsons.copy` to provide a copy facility for our pojos created from json schema. This is because the generated classes do not have a copy method.
* Add `DefaultSchedulerPersistenceTest` - this is main event of this pr. 
* Then a handful of other dead simple test suites for util classes. 
* Working on `JobScheduler` next.

### CODE REVIEW

File: dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.utility.MountableFile;
+
+class DefaultSchedulerPersistenceTest {
+
+  @SuppressWarnings("rawtypes")
+  private static PostgreSQLContainer container;
+  private static BasicDataSource connectionPool;
+
+  private static final SourceConnectionImplementation sourceConnectionImplementation;
```
Comment (cgardens) [inline dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java:72]: @michel-tricot - for the static test objects do you prefer the variable names to be camel cased or constantized e.g. `SOURCE_CONNECTION_IMPLEMENTATION`?

File: dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    @SuppressWarnings("unchecked")
+    Supplier<Instant> timeSupplier = mock(Supplier.class);
+    when(timeSupplier.get()).thenReturn(now);
+
+  }
+
+  @SuppressWarnings("rawtypes")
+  @BeforeAll
+  public static void dbSetup() {
+    container = new PostgreSQLContainer("postgres:13-alpine")
```
Comment (cgardens) [inline dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java:137]: @michel-tricot - i played with the idea of creating a base test case that sets this up (it's identical to some of the tests in dataline-db). i think from what you were saying this morning, you'd prefer to go with the way it's done here and just eat the duplication. is that right?

File: dataline-scheduler/src/main/java/io/dataline/scheduler/SchedulerApp.java [hunk 3]
```diff
@@ -54,8 +54,7 @@
 
   private static final int MAX_WORKERS = 4;
   private static final long JOB_SUBMITTER_DELAY_MILLIS = 1000L;
-  private static final ThreadFactory THREAD_FACTORY =
-      new ThreadFactoryBuilder().setNameFormat("scheduler-%d").build();
+  private static final ThreadFactory THREAD_FACTORY = new ThreadFactoryBuilder().setNameFormat("scheduler-%d").build();
```
Comment (michel-tricot) [inline dataline-scheduler/src/main/java/io/dataline/scheduler/SchedulerApp.java:57]: Not really related to this PR but shoudln't the name be what it is as opposed to what created it? If so then this should be called "worker"

File: dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java (hunk 1 above)
Comment (michel-tricot) [reply dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java:72]: `SOURCE_CONNECTION_IMPLEMENTATION`

File: dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+        connectionPool,
+        ctx -> ctx.fetch("SELECT * FROM jobs WHERE id = ?", jobId).stream()
+            .findFirst()
+            .orElseThrow(
+                () -> new RuntimeException("Could not find job with id: " + jobId)));
+  }
+
+  @Test
+  public void testCreateSourceCheckConnectionJob() throws IOException, SQLException {
+    final SchedulerPersistence schedulerPersistence = new DefaultSchedulerPersistence(connectionPool);
```
Comment (michel-tricot) [inline dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java]: NIT: if you're not going to re-use the persistence what about:
```
final long jobId = new DefaultSchedulerPersistence(connectionPool)
     .createSourceCheckConnectionJob(sourceConnectionImplementation);
```

File: dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    final long jobId = schedulerPersistence.createSourceCheckConnectionJob(sourceConnectionImplementation);
+
+    Record jobEntry = getJobRecord(jobId);
```
Comment (michel-tricot) [inline dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java:185]: everything else is final, why not this one?

File: dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java (hunk 1 above)
Comment (michel-tricot) [reply dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java:72]: Also if some are only used for one test (and there are multiple tests in that testfile), you can create them in the test function directly

File: dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        .withPassword("docker");
+    container.start();
+
+    try {
+      container.copyFileToContainer(MountableFile.forClasspathResource("schema.sql"), "/etc/init.sql");
+      // execInContainer uses Docker's EXEC so it needs to be split up like this
+      container.execInContainer(
+          "psql", "-d", "dataline", "-U", "docker", "-a", "-f", "/etc/init.sql");
+
+    } catch (InterruptedException | IOException e) {
```
Comment (michel-tricot) [inline dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java:150]: don't catch in tests, just throws in the protoype

File: dataline-commons/src/main/java/io/dataline/commons/json/Jsons.java [hunk 7]
```diff
@@ -79,4 +79,9 @@ public static JsonNode deserialize(final String jsonString) {
     return OBJECT_MAPPER.valueToTree(object);
   }
 
+  @SuppressWarnings("unchecked")
```
Comment (michel-tricot) [inline dataline-commons/src/main/java/io/dataline/commons/json/Jsons.java:82]: can you add one test?

**Review by michel-tricot — APPROVED** [review]

File: dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java (hunk 5 above)
Comment (cgardens) [reply dataline-scheduler/src/test/java/io/dataline/scheduler/DefaultSchedulerPersistenceTest.java:185]: bah. no good reason. i need to figure out how to lint this.

**Pull request merged**